### PR TITLE
Integrate libspng

### PR DESCRIPTION
**Pull request by randy408**

### CODE REVIEW

File: projects/libspng/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Copyright 2016 Google Inc.
```
Comment (Dor1s) [inline projects/libspng/Dockerfile]: nit: 2018

**Comment by Dor1s** [comment]
We are ready to accept your project, but there are build issues:

```
+ clang++ -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize-address-use-after-scope -fsanitize=fuzzer-no-link -stdlib=libc++ -std=c++11 -I. /src/libspng/tests/spng_read_fuzzer.cc -o /out/spng_read_fuzzer -lFuzzingEngine /src/libspng/build/libspng.a -lz
/src/libspng/tests/spng_read_fuzzer.cc:1:10: fatal error: '../spng.h' file not found
#include "../spng.h"
         ^~~~~~~~~~~
1 error generated.
Fuzzers build failed.

```

**Comment by randy408** [comment]
Fixed it.

**Comment by Dor1s** [comment]
Thanks. Now there is another problem, a memory leak:

```
Broken fuzz targets (1):
spng_read_fuzzer:
BAD BUILD: /out/spng_read_fuzzer seems to have either startup crash or exit:
INFO: Seed: 2515444350
INFO: Loaded 1 modules   (1063 inline 8-bit counters): 1063 [0x8d1460, 0x8d1887), 
INFO: Loaded 1 PC tables (1063 PCs): 1063 [0x8d1888,0x8d5af8), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2	INITED cov: 19 ft: 20 corp: 1/1b lim: 4 exec/s: 0 rss: 27Mb

=================================================================
==39==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 2280 byte(s) in 1 object(s) allocated from:
    #0 0x4ec837 in calloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:155
    #1 0x5c4fb9 in spng_ctx_new /src/libspng/build/../src/common.c:63:21
    #2 0x5313d1 in LLVMFuzzerTestOneInput /src/libspng/tests/spng_read_fuzzer.cc:5:21
    #3 0x55c6c5 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:571:15
    #4 0x55a597 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /src/libfuzzer/FuzzerLoop.cpp:480:3
    #5 0x55f941 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:766:5
    #6 0x560c78 in fuzzer::Fuzzer::Loop(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:806:3
    #7 0x53e5f5 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:764:6
    #8 0x53162c in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #9 0x7f68ff1a382f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Direct leak of 2280 byte(s) in 1 object(s) allocated from:
    #0 0x4ec837 in calloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:155
    #1 0x5c4fb9 in spng_ctx_new /src/libspng/build/../src/common.c:63:21
    #2 0x5313d1 in LLVMFuzzerTestOneInput /src/libspng/tests/spng_read_fuzzer.cc:5:21
    #3 0x55c6c5 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:571:15
    #4 0x55a597 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /src/libfuzzer/FuzzerLoop.cpp:480:3
    #5 0x55e3a2 in fuzzer::Fuzzer::MutateAndTestOne() /src/libfuzzer/FuzzerLoop.cpp:708:19
    #6 0x561316 in fuzzer::Fuzzer::Loop(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:839:5
    #7 0x53e5f5 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:764:6
    #8 0x53162c in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #9 0x7f68ff1a382f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

Direct leak of 2280 byte(s) in 1 object(s) allocated from:
    #0 0x4ec837 in calloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:155
    #1 0x5c4fb9 in spng_ctx_new /src/libspng/build/../src/common.c:63:21
    #2 0x5313d1 in LLVMFuzzerTestOneInput /src/libspng/tests/spng_read_fuzzer.cc:5:21
    #3 0x55c6c5 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:571:15
    #4 0x55f46b in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:761:3
    #5 0x560c78 in fuzzer::Fuzzer::Loop(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:806:3
    #6 0x53e5f5 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:764:6
    #7 0x53162c in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #8 0x7f68ff1a382f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: 6840 byte(s) leaked in 3 allocation(s).
INFO: to ignore leaks on libFuzzer side use -detect_leaks=0.

MS: 1 ChangeBit-; base unit: adc83b19e793491b1c6ea0fd8b46cd9f32e592fc
0x2a,
*
artifact_prefix='./'; Test unit written to ./leak-df58248c414f342c81e056b40bee12d17a08bf61
Base64: Kg==
ERROR: 100% of fuzz targets seem to be broken. See the list above for a detailed information.
Check build failed.
```

Please follow the New Project Guide and test the PR locally before uploading it for review: https://github.com/google/oss-fuzz/blob/master/docs/new_project_guide.md#testing-locally

**Comment by randy408** [comment]
I have made a few fixes and ran the fuzzer locally for a few minutes with ASan and UBSan without issues. I can't build the coverage report (`cp: cannot stat '/src/honggfuzz/hfuzz_cc/hfuzz-clang': No such file or directory` etc) but it should be around 50% (the testsuite is at 57% currently), I will try to increase this later.

File: projects/libspng/project.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+homepage: "https://libspng.org"
+primary_contact: "randy408@protonmail.com"
+auto_ccs:
+  - "randy440088@gmail.com"
+sanitizers:
```
Comment (Dor1s) [inline projects/libspng/project.yaml]: MSan should work as well (I've tested it locally), I'll remove lines 5-7.

**Comment by kcc** [comment]
@randy408,  quick question: does libspng honor PNG's CRC checks, and if so, 
does the fuzz target bypass them?

**Comment by kcc** [comment]
@randy408, please check https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=12583
the fuzz target is OOM-ing too frequently. 
I'm currently playing with a structure-aware plugin for PNG fuzzers and this fuzz target OOMs almost instantly.

**Comment by randy408** [comment]
@kcc I've [set](https://gitlab.com/randy408/libspng/commit/d8d8ee25f95d5efe059e89165aacc01d146e850f) the output image size limit to ~80MB for spng, for libpng's fuzz target the limit is ~80MB or half of that, not sure, let me know if I should lower it or raise it. Currently there is no API for setting chunk size limits so it will try to allocate up to 2GB per chunk to store it (IDAT chunks are read progressively so those shouldn't cause issues), the relevant issues are [#26](https://gitlab.com/randy408/libspng/issues/26) and [#16](https://gitlab.com/randy408/libspng/issues/16).

**Comment by kcc** [comment]
thanks, seems to have worked. 
Would you be interested in adopting my experimental custom mutator for PNG? 

something like this: 
```
--- a/projects/libspng/Dockerfile
+++ b/projects/libspng/Dockerfile
@@ -23,6 +23,7 @@ RUN apt-get update && \
     apt-get install -y pkg-config zlib1g-dev
 
 RUN git clone --depth 1 https://gitlab.com/randy408/libspng.git
+RUN git clone --depth 1 https://github.com/google/fuzzer-test-suite
 
 WORKDIR libspng
 COPY build.sh $SRC/
diff --git a/projects/libspng/build.sh b/projects/libspng/build.sh
index 62223bb..df1ef57 100644
--- a/projects/libspng/build.sh
+++ b/projects/libspng/build.sh
@@ -24,6 +24,14 @@ $CXX $CXXFLAGS -std=c++11 -I. \
     -o $OUT/spng_read_fuzzer \
     -lFuzzingEngine $SRC/libspng/build/libspng.a -lz
 
+$CXX $CXXFLAGS -std=c++11 -I. \
+    $SRC/libspng/tests/spng_read_fuzzer.cc \
+    -o $OUT/spng_read_fuzzer_structure_aware \
+    -include ../fuzzer-test-suite/libpng-1.2.56/png_mutator.h \
+    -D PNG_MUTATOR_DEFINE_LIBFUZZER_CUSTOM_MUTATOR \
+    -lFuzzingEngine $SRC/libspng/build/libspng.a -lz
+
+


```

I expect you also know about PNG way more than I do, so contribution to the mutator is also welcome! :)

**Pull request merged**